### PR TITLE
feat: Fused temperature online softmax kernel

### PR DESCRIPTION
**Pull request by xslingcn**
## 📌 Description
This adds the fused temperature online softmax kernel to the sampling and logits processor API.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Benchamrk
Baseline: [torch.softmax()](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cuda/SoftMax.cu) on H200
softmax_benchmark_results_32000_128

### CODE REVIEW

File: include/flashinfer/sampling.cuh [hunk 1]
```diff
@@ -261,6 +273,82 @@ __device__ __forceinline__ float GetMaxValue(float* in_data, uint32_t row_idx, u
...
+    if (tx == 0) {
+      temp_storage.shared_state.max_val = block_max;
+    }
+    __syncthreads();
+    block_max = temp_storage.shared_state.max_val;
+
+    float thread_sum = 0.0f;
+#pragma unroll
+    for (uint32_t j = 0; j < VEC_SIZE; ++j) {
+      thread_sum += std::exp(logits_vec[j] - block_max);  // e^(-inf) is safe to add
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh]: use `__expf` instead

File: include/flashinfer/sampling.cuh [hunk 2]
```diff
@@ -261,6 +273,82 @@ __device__ __forceinline__ float GetMaxValue(float* in_data, uint32_t row_idx, u
...
+        cub::BlockReduce<float, BLOCK_THREADS>(temp_storage.block_prim.reduce).Sum(thread_sum);
+    if (tx == 0) {
+      temp_storage.shared_state.denominator = block_sum;
+    }
+    __syncthreads();
+    block_sum = temp_storage.shared_state.denominator;
+
+    if (tx == 0) {
+      float new_max = max(running_max, block_max);
+      running_denominator = running_denominator * std::exp(running_max - new_max) +
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh]: ditto

File: include/flashinfer/sampling.cuh [hunk 3]
```diff
@@ -261,6 +273,82 @@ __device__ __forceinline__ float GetMaxValue(float* in_data, uint32_t row_idx, u
...
+                            block_sum * std::exp(block_max - new_max);
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh]: ditto

File: include/flashinfer/sampling.cuh [hunk 4]
```diff
@@ -261,6 +273,82 @@ __device__ __forceinline__ float GetMaxValue(float* in_data, uint32_t row_idx, u
...
+    running_denominator = temp_storage.shared_state.denominator;
+  }
+
+  const float final_max = running_max;
+  const float inv_denominator = 1.0f / running_denominator;
+
+  __syncthreads();
+
+  // Pass 2: Normalize in place
+  for (uint32_t i = tx; i < d; i += BLOCK_THREADS) {
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh]: set granularity to `VEC_SIZE` instead of 1.

File: include/flashinfer/sampling.cuh [hunk 5]
```diff
@@ -912,6 +1215,92 @@ __global__ void TopKTopPSamplingFromProbKernel(DType* probs, IdType* top_k_arr,
...
+          // Phase 1: Map-Reduce across vocab slices
+          dim3 phase1_nblks(batch_size, num_slices);
+          dim3 phase1_nthrs(BLOCK_THREADS);
+          size_t smem_size = sizeof(OnlineSoftmaxTempStorage<BLOCK_THREADS>);
+
+          auto phase1_kernel = OnlineSoftmaxMapKernel<BLOCK_THREADS, VEC_SIZE, DType>;
+          void* phase1_args[] = {&logits, &partial_results, &temperature_arr, &temperature_val,
+                                 &d,      &num_slices};
+
+          FLASHINFER_CUDA_CALL(cudaFuncSetAttribute(
```
Comment (yzh119) [inline include/flashinfer/sampling.cuh:1275]: Try adding [pdl](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#programmatic-dependent-launch-and-synchronization) support to all of these kernels, a reference can be found at #1117

**Review by yzh119 — APPROVED** [review]
LGTM, in the next PR we should add fusion rules for this pattern.

**Comment by lgeiger** [comment]
Should the calls to `torch.softmax()` in [`top_k_top_p_sampling_from_logits`](https://github.com/flashinfer-ai/flashinfer/blob/9e81467d5489c980e47f9463dd17ffbfca4c4806/flashinfer/sampling.py#L890C5-L890C37) be replaced with this new implementation as well?

https://github.com/flashinfer-ai/flashinfer/blob/9e81467d5489c980e47f9463dd17ffbfca4c4806/flashinfer/sampling.py#L984
https://github.com/flashinfer-ai/flashinfer/blob/9e81467d5489c980e47f9463dd17ffbfca4c4806/flashinfer/sampling.py#L994

**Comment by yzh119** [comment]
Hi @lgeiger good point, would you mind creating a PR for that?

**Comment by byjiang1996** [comment]
Hi @yzh119 @xslingcn , thanks for this awesome PR!

Issue https://github.com/flashinfer-ai/flashinfer/issues/1202: I tried to replace torch.div&torch.softmax to this flashinfer fuse kernel in sglang. [sglang code changes here](https://github.com/byjiang1996/sglang/commit/da5a65e76749e86cc8e74200fd5a08a55c2a7525)

However, in both cases, inference performance degrades after moving to the fuse kernel. Do you have any idea why?

## Before: via pytorch native
model sampler's div&softmax via pytorch native: 45us
image

spec decoding verify's div&softmax via pytorch native: 28us
image

## After: via fuse kernel
model sampler's div&softmax via fuse kernel: 92us
image

spec decoding verify's div&softmax via fuse kernel: 69us
image

## Reproduce
* [sglang code changes here](https://github.com/byjiang1996/sglang/commit/da5a65e76749e86cc8e74200fd5a08a55c2a7525)
* sglang server: `python3 -m sglang.launch_server --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 --attention-backend fa3 --trust-remote-code --dtype float16 --speculative-algo EAGLE3 --model-path /shared/public/models/meta-llama/Meta-Llama-3.1-8B-Instruct --speculative-draft-model-path /shared/public/elr-models/jamesliu1/sglang-EAGLE3-Llama-3.1-Instruct-8B/e5ed08d66f528a95ce89f5d4fd136a28f6def714`
* sglang client: `python3 -m sglang.bench_serving --backend sglang --dataset-name random --num-prompts 16 --sharegpt-output-len 100 --extra-request-body '{"top_k": 4, "temperature": 1.0}' --profile`
